numpy/lib/function_base.py

```diff
@@ -175,54 +175,102 @@ def histogram(a, bins=10, range=None, normed=False, weights=None,
             raise AttributeError(
                 'max must be larger than min in range parameter.')
 
+    # Histogram is an integer or a float array depending on the weights.
+    if weights is None:
+        ntype = int
```

Comment:
Oh, I see! It then is a bug on the current implementation. An unlikely bug, but the histogram of an array of `2**31` items on a 64 bit system, all of which end up in the same bin, will show a `-1` count for that bin, which is not quite what we are after. Since you are at it, you may as well fix it, since it does affect your code: I'm pretty sure `bincount` always returns either `np.intp` (without weights) or `np.double` (with weights).